graph/toposort.cxx

```diff
@@ -0,0 +1,73 @@
+#include <iostream>
+#include <set>
+#include <map>
+#include <queue>
+#include <vector>
+
+template <typename T>
+class Graph {
+	std::map<T, std::set<T>> g;
+public:
+	void addEdge(T a, T b) {
+		g[a].insert(b);
+	}
+	void removeEdge(T a, T b) {
+		g[a].insert(b);
+	}
+
+	std::vector<T> toposort() {
+		std::set<T> visited;
+		std::vector<T> result;
+
+		for(auto const& v : g) {
+			if(visited.find(v.first) == visited.end())
+				dfs(visited, v.first, result);
+		}
+		return result;
+	}
+	void dfs(std::set<T>& visited, const T& v, std::vector<T>& result) {
+		visited.insert(v);
+		for(auto const& e : g[v]) {
+			if(visited.find(e) == visited.end()) {
+				dfs(visited, e, result);
+			}
+ 		}
+		result.push_back(v);
+	}
+
+	void bfs(graph<T> g, T start) {
```

Comment:
Its a little weird that this is a member function of the graph class, but also takes a graph as the argument?  Also, what is the use case of this function?